File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size();
        vector<int>pref(n + 1);
        
        for (int i = 1; i <= n; i++) {
            pref[i] = pref[i - 1] + (nums[i - 1] == 0);
        }
        
        int start = 0, ans = 0;
        for (int end = 1; end <= n; end++) {
            while (pref[end] - pref[start] > k)
                start++;
            ans = max(ans, end - start);
        }
        return ans;
    }
};

File: 1004-max-consecutive-ones-iii/1004-max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int n = nums.length;
        int[] pref = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            pref[i] = pref[i - 1];
            if (nums[i - 1] == 0)
                pref[i]++;
        }
        
        int left = 0, ans = 0;
        int right = 0;
        while (right <= n) {
            while (pref[right] - pref[left] > k)
                left++;
            ans = Math.max(ans, right - left);
            right++;
        }
        return ans;
    }
}